### ISSUE TITLE
Object that are not entities are checked for Metadata without cache

### ISSUE THREAD

**Issue by bastnic** [doctrine/persistence#96]
This is a followup of https://github.com/doctrine/dbal/issues/3838, I just want to check if maintainers see a room for improvement.

TL;DR: when we check metadata for doctrine entities, there is a cache that prevent checking for metadata across requests and there is memoization to prevent hiting on the cache. When passing object that are not doctrine entities with fallback there is no cache but memoization, and without fallback there is none of that. 

I don't have (yet) a blackfire profile on that, but I guess on certain scenario it can be ugly.

There is at least two path that trigger that comportment (just to show that this is not an easy fix in userland):

1/ doctrine query with datetime parameter without the third parameter "datetime"

- https://github.com/doctrine/orm/blob/master/lib/Doctrine/ORM/AbstractQuery.php#L402
- https://github.com/doctrine/orm/blob/master/lib/Doctrine/ORM/UnitOfWork.php#L2491
- https://github.com/doctrine/orm/blob/master/lib/Doctrine/ORM/EntityManager.php#L286
- https://github.com/doctrine/persistence/blob/master/lib/Doctrine/Persistence/Mapping/AbstractClassMetadataFactory.php#L161
- https://github.com/doctrine/persistence/blob/master/lib/Doctrine/Persistence/Mapping/AbstractClassMetadataFactory.php#L206

This is fixable in userland, but quite not fun.

2/ symfony validation

- https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php#L101
- https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Validator/Mapping/Loader/LoaderChain.php#L54
- https://github.com/symfony/symfony/blob/master/src/Symfony/Bridge/Doctrine/Validator/DoctrineLoader.php#L51
- https://github.com/doctrine/orm/blob/master/lib/Doctrine/ORM/EntityManager.php#L286
- https://github.com/doctrine/persistence/blob/master/lib/Doctrine/Persistence/Mapping/AbstractClassMetadataFactory.php#L161
- https://github.com/doctrine/persistence/blob/master/lib/Doctrine/Persistence/Mapping/AbstractClassMetadataFactory.php#L206

---- 

I don't know the best fix for that but I figures that on the life on a request / cache, this should not happens. Even I which know this bug, I forgot to add the third parameter, it should really be resolved in doctrine/persistence.

I think I would cache the exception and add some instanceof / throw somewhere.

**Comment by greg0ire** [comment]
IMO the most BC way would be to add a new array, and cache exceptions in that array, that way you are sure to throw the same type of exception as before. The issue is that it will be an exception with an outdated stack trace. It's an issue but ideally (in the future), that code path should not even be used by apps. There should be a new method (that would have been named `hasMetadataFor`, but will not because that name is already taken for something that seems completely useless) that returns a boolean and that you would be supposed to check before calling `loadMetadataFor`, because relying on exceptions for things that are not exceptional does not make a lot of sense. Since `loadedMetadata` is exposed through `getLoadedMetadata`, touching that array is a no-go, even though it is private and even though that method does not seem to relied on much.
